parlai/agents/bert_qa/bert_qa.py

```diff
@@ -0,0 +1,443 @@
+                segment_ids.append(0)
+
+            b_input_mask.append(input_mask)
+
+            assert len(tokens_ids) == max_tokens_length
+            assert len(input_mask) == max_tokens_length
+            assert len(segment_ids) == max_tokens_length
+
+        return (
+            torch.tensor(b_tokens_ids, dtype=torch.long).to(self.device),
+            torch.tensor(b_segment_ids, dtype=torch.long).to(self.device),
+            torch.tensor(b_input_mask, dtype=torch.long).to(self.device),
+            torch.tensor(b_start_position, dtype=torch.long).to(self.device),
+            torch.tensor(b_end_position, dtype=torch.long).to(self.device),
+        )
+
+    def _get_best_indexes(self, logits, n_best_size):
+        """Get the n-best logits from a list."""
+        index_and_score = sorted(enumerate(logits), key=lambda x: x[1], reverse=True)
+
+        best_indexes = []
+        for i in range(len(index_and_score)):
+            if i >= n_best_size:
+                break
+            best_indexes.append(index_and_score[i][0])
+        return best_indexes
+
+    def _get_prediction(
+        self, start_logits, end_logits, input_ids, start_context_position=None
+    ):
+
+        start_indexes = self._get_best_indexes(start_logits, self.opt["n_best_size"])
+        end_indexes = self._get_best_indexes(end_logits, self.opt["n_best_size"])
+
+        _PrelimPrediction = collections.namedtuple(
+            "PrelimPrediction", ["start_index", "end_index", "start_logit", "end_logit"]
+        )
+        prelim_predictions = []
+
+        for start_index in start_indexes:
+            for end_index in end_indexes:
+                # We could hypothetically create invalid predictions
+                if end_index < start_index:
+                    continue
+                if start_context_position and start_index < start_context_position:
+                    # answer in the question
+                    continue
+                length = end_index - start_index + 1
+                if length > self.text_truncate:
+                    continue
+                prelim_predictions.append(
+                    _PrelimPrediction(
+                        start_index=start_index,
+                        end_index=end_index,
+                        start_logit=start_logits[start_index],
+                        end_logit=end_logits[end_index],
+                    )
+                )
+
+        prelim_predictions = sorted(
+            prelim_predictions,
+            key=lambda x: (x.start_logit + x.end_logit),
+            reverse=True,
+        )
+
+        _NbestPrediction = collections.namedtuple(  # pylint: disable=invalid-name
+            "NbestPrediction", ["text", "start_logit", "end_logit"]
+        )
+
+        nbest = []
+        for pred in prelim_predictions:
+            if len(nbest) >= self.opt["n_best_size"]:
+                break
+
+            if pred.start_index > 0:  # this is a non-null prediction
```

Comment:
Can it be <=0 anyway, given l.216?

Reply:
mmmm, I don't think so. The first token is always [CLS], so a valid start_index should be >= 1